crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -112,3 +106,167 @@ impl Pruner {
+    watermark: &WatermarkRead,
+    cancel: &CancellationToken,
+    delay_amount: i64,
+) -> IndexerResult<()> {
+    let current_time = chrono::Utc::now().timestamp_millis();
+    let delay = (watermark.timestamp_ms + delay_amount - current_time).max(0) as u64;
+
+    if delay > 0 {
+        info!("Waiting for {}ms before pruning", delay);
+        tokio::select! {
+            _ = tokio::time::sleep(Duration::from_millis(delay)) => Ok(()),
+            _ = cancel.cancelled() => {
+                info!("Pruning cancelled during delay");
+                Ok(())
+            }
+        }
+    } else {
+        Ok(())
+    }
+}
+
+/// Fetches all entries from the `watermarks` table, and updates the lower bounds for all watermarks
+/// if the entry's epoch range exceeds the respective retention policy.
+async fn update_watermarks_lower_bounds(
+    store: &PgIndexerStore,
+    epochs_to_keep: u64,
+) -> IndexerResult<()> {
+    let watermarks = store.get_watermarks().await?;
+    let mut lower_bound_updates = vec![];
+
+    for (key, value) in watermarks.iter() {
+        let epochs_to_keep = if key.as_str() == "objects_history" {
+            OBJECTS_HISTORY_EPOCHS_TO_KEEP
+        } else {
+            epochs_to_keep
+        };
+
+        // We should update the watermarks and prepare for pruning
+        if should_update_watermark(value, epochs_to_keep) {
+            let new_inclusive_epoch_lower_bound = value.epoch_hi.saturating_sub(epochs_to_keep - 1);
+            // TODO: (wlmyng) don't rely on `checkpoints` table
+            let (cp, tx) = store
+                .get_min_cp_and_tx_for_epoch(new_inclusive_epoch_lower_bound)
+                .await?;
+            let new_lo = match key {
+                WatermarkEntity::ObjectsHistory | WatermarkEntity::Checkpoints => cp,
+                WatermarkEntity::Transactions | WatermarkEntity::Events => tx,
+                WatermarkEntity::Epochs => new_inclusive_epoch_lower_bound,
+            };
+            lower_bound_updates.push(Watermark::lower_bound(
+                *key,
+                new_inclusive_epoch_lower_bound,
+                new_lo,
+            ));
+        }
+    }
+
+    if !lower_bound_updates.is_empty() {
+        store.update_watermarks(lower_bound_updates).await?;
+        info!("Finished updating lower bounds for watermarks");
+    }
+
+    Ok(())
+}
+
+/// Task to periodically query the `watermarks` table and update the lower bounds for all watermarks
+/// if the entry exceeds epoch-level retention policy.
+async fn update_watermarks_lower_bounds_task(
+    store: PgIndexerStore,
+    epochs_to_keep: u64,
+    cancel: CancellationToken,
+) -> IndexerResult<()> {
+    loop {
+        if cancel.is_cancelled() {
+            info!("Pruner watermark lower bound update task cancelled.");
+            return Ok(());
+        }
+
+        update_watermarks_lower_bounds(&store, epochs_to_keep).await?;
+        tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
nit: use an [`Interval`](https://docs.rs/tokio/latest/tokio/time/struct.Interval.html) to avoid the tick rate skewing because of the amount of time the operation takes.

Reply:
Also, `select` on the `sleep` and the cancellation token so that you don't have to wait up to 5 seconds after a cancellation request to honour it.